### PR TITLE
Refactor/memory tutorial

### PR DESCRIPTION
**Pull request by mbartling**
Significant refactoring to memory tutorials. 

Marked for tutorials:
* `memory_intro.md`
* `runtime_stats.md`
* `runtime_mem_trace.md`

Marked for move to reference:
* everything else

### CODE REVIEW

File: docs/tutorials/optimize/memory/memap.md [hunk 1]
```diff
@@ -1,10 +1,10 @@
-### memap - static memory map analysis
+## memap - static memory map analysis
```
Comment (AnotherButler) [inline docs/tutorials/optimize/memory/memap.md]: I don't want to change heading levels until after we move pages. Please fix all heading levels to how they were before.

File: docs/tutorials/optimize/memory/memory_intro.md [hunk 2]
```diff
@@ -1,56 +1,5 @@
-## Memory Optimization
+## Memory optimizations
```
Comment (AnotherButler) [inline docs/tutorials/optimize/memory/memory_intro.md]: This page is now very short and doesn't seem to add a lot of content.

File: docs/tutorials/optimize/memory/runtime_stats.md [hunk 3]
```diff
@@ -1,25 +1,53 @@
...
-* Use `alloc_fail_cnt` to check if allocations have been failing, and if so, how many.
+1. Use the function `mbed_stats_heap_get()` to take a snapshot of heap stats.
 
-##### Example program using heap statistics
+   <span class="notes">**Note**: This function is available even when the heap stats are not enabled but always returns zero for all fields.</span>
+
+   `mbed_stats_heap_get()` returns a struct containing the following:
+
+
+   |                   |                                                   |
```
Comment (AnotherButler) [inline docs/tutorials/optimize/memory/runtime_stats.md]: Query: What is the empty extra row for? Can we remove it?

File: docs/tutorials/optimize/memory/runtime_stats.md [hunk 4]
```diff
@@ -46,31 +74,53 @@ int main(void)
...
+- `mbed_stats_stack_get` calculates combined stack informations for all threads.
+- `mbed_stats_stack_get_each` provides stack informations for each thread separately.
 
 <span class="notes">**Note**: These functions are available even when the stack stats are not enabled but always return zero for all fields.</span>
 
-#### Example use cases
+Both of these functions return a struct containing the following:
+
+
+|                 |                                                                           |
```
Comment (AnotherButler) [inline docs/tutorials/optimize/memory/runtime_stats.md]: Query: What is the empty extra row for? Can we remove it?

File: docs/tutorials/optimize/memory/static_memory_optimization.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+##### Embedded targets
+
+You can also take advantage of the fact that these programs only run on embedded targets. When you run a C++ application on a desktop computer, the operating system constructs every global C++ object before calling `main`. It also registers a handle to destroy these objects when the program ends. The code the compiler injects has some implications for the application:
+
+- The code that the compiler injects consumes memory.
+- It implies dynamic memory allocation and thus requires the binary to include `malloc`, even when the application does not use it.
+
+When you run an application on an embedded device, you don't need handlers to destroy objects when the program exits, because the application will never end. You can save more RAM and flash memory usage by [removing destructor registration](https://github.com/ARMmbed/mbed-os/pull/2745) on application startup and by eliminating the code to destruct objects when the operating system calls `exit()` at runtime. 
+
+<span class="notes">**Note:** Mbed OS 5.2.0 enables removing destructor registration by default.</span>
```
Comment (AnotherButler) [inline docs/tutorials/optimize/memory/static_memory_optimization.md]: Note: I updated this for active voice and formatting. Please make sure I didn't accidentally change the content.

File: docs/tutorials/optimize/memory/memory_intro.md (hunk 2 above)
Comment (mbartling) [reply docs/tutorials/optimize/memory/memory_intro.md]: This is going to be expanded, it just wasn't in my jira ticket

File: docs/tutorials/optimize/memory/runtime_stats.md (hunk 3 above)
Comment (mbartling) [reply docs/tutorials/optimize/memory/runtime_stats.md]: Tables don't render properly without that space

File: docs/tutorials/optimize/memory/runtime_stats.md (hunk 4 above)
Comment (mbartling) [reply docs/tutorials/optimize/memory/runtime_stats.md]: Tables won't render without it. Verified with the docs engine

File: docs/tutorials/optimize/memory/static_memory_optimization.md (hunk 5 above)
Comment (AnotherButler) [reply docs/tutorials/optimize/memory/static_memory_optimization.md]: Please add + after 5.2.0

**Pull request merged**